# Chapter 2

# Types of Data

### 2.1 Strings

In [1]:
x = "Hello World"

In [2]:
print(x)

Hello World


Nhập dữ liệu x là "Hello World" và in dữ liệu x với kết quả trả về là Hello World.

In [3]:
y = 'How are you?'
print(y)

How are you?


Sử dụng dấu nháy đơn 'How are you?' để tạo một chuỗi và gán chuỗi đó vào biến y với kết quả trả về là How are you?

In [4]:
z = 'I\'m fine'
print(z)

I'm fine


Khi sử dụng dấu nháy đơn '...' để tạo chuỗi, nếu bên trong chuỗi cũng có dấu nháy đơn ví dụ như 'I'm, thì phải dùng ký tự thoát \ để tránh lỗi cú pháp. Cuối cùng trả về kết quả biến y là I'm fine.

In [1]:
z = "I'm \"fine\""
print(z)

I'm "fine"


Khi chuỗi được tạo bằng dấu nháy kép "...", có thể dùng dấu nháy đơn ' bên trong mà không cần escape. Tuy nhiên, nếu muốn chèn dấu nháy kép bên trong chuỗi, ta cần dùng ký tự thoát \".

In [2]:
x = "Hello, \nWorld"
print(x)

Hello, 
World


Ký tự \n được dùng để tạo dòng mới trong chuỗi. Khi in ra bằng print(), nội dung sau ký tự \n sẽ được hiển thị trên dòng tiếp theo.

In [3]:
x = """Hello,
World"""
print(x)

Hello,
World


Ba dấu nháy kép """ ... """ cho phép tạo chuỗi nhiều dòng nên khi xuống dòng trong cùng nội dung chuỗi sẽ được giữ nguyên khi in ra.

In [4]:
x = "Hello World !"
print(len(x))

13


Hàm len() được dùng để đếm số ký tự trong chuỗi, bao gồm cả khoảng trắng và dấu câu nên kết quả print trả về là 13.

In [5]:
print(x, len(x))

Hello World ! 13


Lệnh print(x, len(x)) in ra cả nội dung chuỗi x và số ký tự trong chuỗi đó, kết quả trả về là Hello World ! 13 cách nhau bằng một khoảng trắng.

#### 2.1.1 Concatenation of Strings

In [6]:
print("Hello" + " World")

Hello World


Phép cộng (+) dùng để nối hai chuỗi lại với nhau thành một chuỗi mới. 

In [7]:
print( 3 * "Go Habs Go! " + "Woo Hoo!")

Go Habs Go! Go Habs Go! Go Habs Go! Woo Hoo!


Phép toán * khi áp dụng với chuỗi và số nguyên sẽ tạo ra chuỗi lặp lại số lần tương ứng.

In [8]:
x = ('You shall ' 'not ' "pass!")
print(x)

You shall not pass!


Khi đặt nhiều chuỗi liên tiếp liền nhau trong ngoặc đơn, Python sẽ tự động nối chúng thành một chuỗi duy nhất mà không cần dùng đến phép toán +.

In [9]:
x = "I like to code in {}"
langage_1 = "R"
langage_2 = "Python"
preference_1 = x.format(langage_1)
print(preference_1)

I like to code in R


Chuỗi chứa một dấu ngoặc nhọn {} là vị trí chứa chỗ trống để chèn giá trị. Khai báo 2 biến chuỗi với language_1 là R và langugage_ 2 là Python. Gọi lệnh .format() trên chuỗi x với giá trị của langage_1 là R sẽ được chèn vào vị trí {} trong chuỗi x. Kết quả in ra màn hình sẽ là I like to code in R.

In [2]:
x = "I like to code in {}"
langage_1 = "R"
langage_2 = "Python"
preference_2 = x.format(langage_2)
print(preference_2)

I like to code in Python


Chuỗi chứa một dấu ngoặc nhọn {} là vị trí chứa chỗ trống để chèn giá trị. Khai báo 2 biến chuỗi với language_1 là R và langugage_ 2 là Python. Gọi lệnh .format() trên chuỗi x với giá trị của langage_2 là Python sẽ được chèn vào vị trí {} trong chuỗi x. Kết quả in ra màn hình sẽ là I like to code in Python.

In [3]:
x = "I like to code in {} and in {}"
preference_3 = x.format(langage_1, langage_2)
print(preference_3)

I like to code in R and in Python


Khi một chuỗi chứa nhiều dấu {}, ta có thể chèn nhiều giá trị tương ứng bằng cách truyền các biến vào format() theo đúng thứ tự. Kết quả in ra màn hình sẽ đúng theo giá trị tương ứng.

#### 2.1.2 Indexing and Extraction

In [4]:
x[i-1]

NameError: name 'i' is not defined

Chuỗi trong Python có thể được truy cập từng ký tự bằng chỉ số. Ký tự đầu tiên có chỉ số là 0, ký tự thứ i có chỉ số là i-1. Dùng cú pháp x[i-1] để lấy ký tự đó nhưng hiện thông báo lỗi NamError nên phải khai báo biến i trước khi sử dụng trong biểu thức x[i-1].

In [5]:
x = "Hello"
print(x[0])

H


Đây là cú pháp truy cập ký tự tại vị trí thứ 0 trong chuỗi x và trong Python, chỉ số bắt đầu từ 0, nên print(x[0]) thì ký tự đầu tiên của chuỗi "Hello" sẽ trả về kết quả in ra màn hình là H.



In [7]:
print(x[4])

o


Tương tự, ký tự thứ 4 của chuỗi "Hello" sẽ trả về kết quả in ra màn hình là o.

In [8]:
print(x[-2])

l


Chỉ số ký tự âm sẽ được đánh từ bên phải sang trái, bắt đầu từ -1, nên ký tự -2 sẽ in ra là 

In [1]:
x = "You shall not pass!"
# From the fourth character (not included) to the ninth (included)
print(x[4:9])

shall


Chuỗi x có nội dung là: "You shall not pass!" và lệnh print(x[4:9]) sẽ lấy các ký tự từ vị trí 4 đến trước vị trí 9 (không bao gồm vị trí 9) nên kết quả trả về là shall.

In [2]:
# From the 4th character (non included) to the end of the string
print(x[4:])
# From the beginning of the string to the penultimate (included)
print(x[:-1])
# From the 3rd character before the end (included) to the end
print(x[-5:])

shall not pass!
You shall not pass
pass!


Lệnh print(x[4:]) sẽ lấy ký tự từ chỉ số 4 đến hết chuỗi, lệnh print(x[:-1]) lấy từ đầu chuỗi đến trước ký tự cuối cùng - tức là bỏ ký tự cuối. Cuối cùng, lệnh print(x[-5:]) lấy chuỗi bắt đầu theo thứ tự chỉ số âm từ thứ 5 từ cuối đến hết.

In [3]:
# From the 4th character (not included),
# to the end of the string, in steps of 3
print(x[4::3])

sln s


Lệnh print(x[4::3]) sẽ in ra các ký tự bắt đầu từ vị trí thứ 4 của chuỗi (ký tự "s") và lấy tiếp mỗi 3 ký tự một lần, cho đến khi hết chuỗi. Kết quả trả về sẽ là sln s.


In [2]:
x = "You shall not pass!"
print(x[::-1])

!ssap ton llahs uoY


Lệnh print(x[::-1]) sẽ in ra chuỗi x theo thứ tự ngược lại.

#### 2.1.3 Available Methods with Strings

In [3]:
x = "le train de tes injures roule sur le rail de mon indifférence"
print(x.count("in"))

3


Lệnh print(x.count("in")) sẽ đếm số lần chuỗi con "in" xuất hiện trong chuỗi x và in kết quả ra màn hình.

##### 2.1.3.1 Coversion to upper or lower case

In [6]:
x = "le train de tes injures roule sur le rail de mon indifférence"
print(x.lower())
print(x.upper())

le train de tes injures roule sur le rail de mon indifférence
LE TRAIN DE TES INJURES ROULE SUR LE RAIL DE MON INDIFFÉRENCE


x.lower() → Trả về bản sao của chuỗi x với toàn bộ ký tự chuyển thành chữ thường.

x.upper() → Trả về bản sao của chuỗi x với toàn bộ ký tự chuyển thành chữ in hoa.

##### 2.1.3.2 Seach Pattern for Strings

In [3]:
print(x.find("in"))
print(x.find("hello"))

6
-1


Lệnh x.find("in") sẽ tìm vị trí (chỉ số nhỏ nhất) của chuỗi con "in" trong x. nếu tìm thấy, trả về chỉ số đó.

x.find("hello") → Tìm vị trí của chuỗi con "hello" trong x. Nếu không tìm thấy, trả về -1.

In [5]:
print(x.find("in", 7, 20))

16


Lệnh yêu cầu tìm chuỗi con "in" trong x nhưng chỉ tìm trong phạm vi từ chỉ số 7 đến chỉ số 20, kết quả trả về màn hình là chỉ số 16

In [7]:
print(x.find("in", 20))

49


Tương tự, lệnh yêu cầu tìm chuỗi con "in" trong x, bắt đầu từ vị trí 20 đến hết chuỗi.

In [8]:
x = "Mademoiselle Deray, il est interdit de manger de la choucroute ici."
print(x.find("deray"))

-1


Lệnh minh họa việc tìm chuỗi con nhưng có phân biệt chữ hoa và chữ thường. x.find("deray") trả về kết quả -1 vì "deray" (chữ thường) không khớp với "Deray" trong chuỗi x.

In [9]:
print(x.capitalize().find("deray"))

13


Chuyển chuỗi x thành dạng viết hoa chữ cái đầu, còn lại viết thường, rồi tìm vị trí xuất hiện của "deray" trong kết quả đó và in ra vị trí này.

##### 2.1.3.3 Splitting Strings

In [10]:
print(x.split(" "))

['Mademoiselle', 'Deray,', 'il', 'est', 'interdit', 'de', 'manger', 'de', 'la', 'choucroute', 'ici.']


Tách chuỗi x thành một danh sách các chuỗi con, dùng dấu cách " " làm ký tự phân tách, rồi in kết quả danh sách đó ra màn hình.

In [3]:
# Will return the elements matched up to the index 3
print(x.split(" ", 3))

['Mademoiselle', 'Deray,', 'il', 'est interdit de manger de la choucroute ici.']


Tách chuỗi x thành tối đa 4 phần bao gồm 3 lần tách dựa trên dấu cách " ", rồi in kết quả danh sách các phần đó ra màn hình.

In [4]:
x = '''"No, I am your Father!- No... No. It's not true! That's impossible!- Search your feelings. You know it to be true.- Noooooooo! Noooo!"'''
print(x.splitlines())

['"No, I am your Father!- No... No. It\'s not true! That\'s impossible!- Search your feelings. You know it to be true.- Noooooooo! Noooo!"']


Tách chuỗi x thành một danh sách các dòng dựa trên ký tự xuống dòng (\n, \r\n...), rồi in kết quả danh sách đó ra màn hình.

##### 2.1.3.4 Cleaning, completion

In [6]:
x = "\n\n    Pardon, du sucre ?    \n \n"
print(x.strip())

Pardon, du sucre ?


Xóa tất cả các ký tự trống (như khoảng trắng, xuống dòng \n, tab \t…) ở đầu và cuối chuỗi x, rồi in kết quả phần nội dung hoàn chỉnh còn lại ra màn hình.

In [7]:
x = "www.egallic.fr"
print(x.strip("wrf."))

egallic


Xóa tất cả các ký tự thuộc nhóm "w", "r", "f", "." ở đầu và cuối chuỗi x, rồi in kết quả phần nội dung hoàn chỉnh còn lại ra màn hình.

In [8]:
longitude = "48.11"
print(longitude.rjust(7," "))

  48.11


Lệnh longitude yêu cầu căn phải trong một chuỗi có tổng độ dài 7 ký tự, thêm ký tự khoảng trắng " " vào bên trái nếu cần, rồi in kết quả ra màn hình.

##### 2.1.3.5 Replacements

In [9]:
x = "Criquette ! Vous, ici ? Dans votre propre salle de bain ? Quelle surprise !"
print(x.replace("Criquette", "Ridge"))

Ridge ! Vous, ici ? Dans votre propre salle de bain ? Quelle surprise !


Thay thế tất cả các lần xuất hiện của chuỗi "Criquette" trong x bằng chuỗi "Ridge", rồi in kết quả ra màn hình.

In [10]:
print(x.replace(" ", ""))

Criquette!Vous,ici?Dansvotrepropresalledebain?Quellesurprise!


Xóa tất cả dấu cách " " trong chuỗi x bằng cách thay chúng bằng chuỗi rỗng "", rồi in kết quả ra màn hình.

#### 2.1.4 Conversion to character strings

In [12]:
nb_followers = 0
message = "He has " + nb_followers + "followers."

TypeError: can only concatenate str (not "int") to str

Lệnh đang cố gắng nối một chuỗi với một số nguyên, điều mà Python không cho phép trực tiếp nên gây ra lỗi TypeError.

In [13]:
print(message)

NameError: name 'message' is not defined

Biến message chưa được tạo hoặc chưa gán giá trị trước khi gọi print nên Python báo lỗi NameError.

In [14]:
message = "He has " + str(nb_followers) + " followers."
print(message)

He has 0 followers.


Chuyển giá trị số nb_followers sang chuỗi bằng str(), sau đó nối với các chuỗi còn lại để tạo câu hoàn chỉnh, rồi in kết quả câu đó ra màn hình.

##### 2.1.5 Exercise

1. Create two variables named a and b so that they contain the following strings respectively: 23 to 0 and C'est la piquette, Jack!.

In [2]:
a = "23 to 0"
b = "C'est la piquette, Jack!"

2. Display the number of characters from a, then b.

In [18]:
print(len(a))  
print(len(b))

7
24


3. Concatenate a and b in a single string, adding a comma as a separating character.

In [19]:
c = a + ", " + b
print(c)

23 to 0, C'est la piquette, Jack!


4. Same question by choosing the separation line as the separator character.

In [20]:
c = a + "\n" + b
print(c)

23 to 0
C'est la piquette, Jack!


5. Using the appropriate method, capitalize a and b.

In [21]:
print(a.capitalize())
print(b.capitalize())

23 to 0
C'est la piquette, jack!


6. Using the appropriate method, lowercase a and b.

In [22]:
print(a.lower())
print(b.lower())

23 to 0
c'est la piquette, jack!


7. Extract the word la and Jack from the string b, using indexes.

In [24]:
word1 = b[6:8]   
word2 = b[19:24]  
print(word1)
print(word2)

la
Jack!


8. Look for the sub-chain piqu in b, then do the same with the sub-chain mauvais.

In [25]:
print(b.find("piqu"))     
print(b.find("mauvais"))

9
-1


9. Return the position (index) of the first character a found in the string b, then try with the character w.

In [3]:
pos_a = b.find("a")  
pos_w = b.find("w")  
print(pos_a)  
print(pos_w)  

7
-1


10. Replace the occurrences of the pattern a by the pattern Z in the substring b.

In [4]:
b_new = b.replace("a", "Z")
print(b_new)

C'est lZ piquette, JZck!


11. Separate the string b using the comma as a sub-chain separator.

In [5]:
parts = b.split(",")
print(parts)

["C'est la piquette", ' Jack!']


12. (Bonus) Remove all punctuation characters from string b, then use an appropriate method to remove white characters at the beginning and end of the string. (Use the ‘regex’ library).

In [6]:
import re
b = "C'est la piquette, Jack!"
# Xóa dấu câu
b_no_punct = re.sub(r"[^\w\s]", "", b)
# Xóa khoảng trắng đầu/cuối
b_clean = b_no_punct.strip()
print(b_clean)

Cest la piquette Jack


### 2.2 Numerical values

#### 2.2.1 Integers

In [3]:
x= 2
y = -2 
print(type(x))

<class 'int'>


Tạo biến x bằng 2 và y bằng -2, sau đó trả kết quả kiểu dữ liệu của x là số nguyên int.

In [4]:
print(type(y))

<class 'int'>


Gán giá trị số nguyên âm -2 cho biến y, sau đó in ra kiểu dữ liệu của y. Kết quả trả về cho biết y thuộc kiểu int (số nguyên).

#### 2.2.2 Floating Point Numbers

In [5]:
x = 2.0
y = 48.15162342 
print(type(x))

<class 'float'>


Tạo biến x với giá trị 2.0 và biến y với giá trị 48.15162342, sau đó in ra kiểu dữ liệu của x. Kết quả cho biết x thuộc kiểu float (số thực).

In [6]:
print(type(y))

<class 'float'>


Tương tự với biến y, kết quả trả về biết y thuộc kiểu số thực (float)

In [7]:
x = 3.2E12 
y = 3.2e12 
print(x)

3200000000000.0


Gán cho x và y giá trị theo dạng ký hiệu khoa học 3.2×10^12 (float), sau đó in giá trị của x. Kết quả in ra là 3200000000000.0.

In [8]:
print(y)

3200000000000.0


Tương tự với biến y.

In [10]:
print(0.35)

0.35


0.35 là số thực nên kết quả trả về sẽ là 0.35 (float)

In [11]:
print(.35)

0.35


In ra màn hình giá trị số thực 0.35 được viết ở dạng rút gọn .35

#### 2.2.3 Complex numbers

In [12]:
z = 1+3j 
print(z)

(1+3j)


Tạo một biến z là số phức với phần thực bằng 1 và phần ảo bằng 3, sau đó in ra màn hình. Kết quả in ra là (1+3j).

In [13]:
print(type(z))

<class 'complex'>


In ra kiểu dữ liệu của biến z. Kết quả cho biết z thuộc kiểu complex (số phức).

In [14]:
z = complex(1, 3) 
print(z)

(1+3j)


Tạo số phức với phần thực bằng 1 và phần ảo bằng 3 bằng hàm complex(), sau đó in ra màn hình. Kết quả in ra là (1+3j).

In [15]:
print(type(z))

<class 'complex'>


In ra kiểu dữ liệu của biến z. Kết quả cho biết z thuộc kiểu complex (số phức).

In [16]:
print(z.conjugate())

(1-3j)


Lấy số phức liên hợp của z và in ra. Kết quả hiển thị là (1-3j)

In [17]:
z = complex(1, 3) 
print(z.real)

1.0


Tạo số phức với phần thực bằng 1 và phần ảo bằng 3, sau đó in ra giá trị phần thực (1.0).

In [19]:
print(z.imag)

3.0


Lấy và in ra phần ảo của số phức z. Kết quả là 3.0.

#### 2.2.4 Conversions

##### 2.2.4.1 Conversion to Integer

In [20]:
x = "3"
x_int = int(x) 
print(type(x_int))

<class 'int'>


Chuyển đổi giá trị chuỗi "3" thành số nguyên bằng hàm int(), sau đó in kết quả ra màn hình kiểu dữ liệu của kết quả là int.

In [21]:
print(type(x))

<class 'str'>


In ra kiểu dữ liệu của biến x. Kết quả cho biết x thuộc kiểu str (chuỗi ký tự).

In [22]:
x = 3.6
x_int = int(x) 
print(x_int)

3


Gán số thực 3.6 cho biến x, sau đó chuyển đổi sang số nguyên (cắt phần thập phân) và in ra màn hình kết quả là 3.

##### 2.2.4.2 Conversion to Floating Point Number 

In [23]:
x = "3.6"
x_float = float(x) 
print(type(x_float))

<class 'float'>


Gán chuỗi "3.6" cho biến x. Dùng hàm float() để chuyển giá trị của x sang số thực và gán cho biến x_float và in ra kiểu dữ liệu của x_float, kết quả là <class 'float'>.

In [25]:
x = 3
x_float = float(x) 
print(x_float)

3.0


Gán giá trị số nguyên 3 cho biến x. Dùng hàm float() để chuyển giá trị của x sang số thực và gán cho biến x_float và in ra giá trị của x_float, kết quả là 3.0.

##### 2.2.4.3 Conversion to Complex

In [26]:
x = "2"
x_complex = complex(x) 
print(x_complex)

(2+0j)


Dùng hàm complex() để chuyển giá trị của x sang số phức và gán cho biến x_complex. In ra giá trị của x_complex, kết quả là (2+0j).

In [28]:
x = 2.4
x_complex = complex(x) 
print(x_complex)

(2.4+0j)


Gán giá trị số thực 2.4 cho biến x. Dùng hàm complex() để chuyển giá trị của x sang số phức và gán cho biến x_complex. In ra giá trị của x_complex, kết quả là (2.4+0j).

### 2.3 Booleans

In [2]:
x = True
y = False 
print(x, y)

True False


Gán giá trị logic True cho biến x và giá trị logic False cho biến y. In ra giá trị của x và y, kết quả là True False.

In [3]:
res = True + True + False + True*True 
print(res)

3


Giá trị logic True và False khi tham gia phép toán sẽ được chuyển thành số nguyên 1 và 0 tương ứng. Câu lệnh trên thực hiện phép cộng và nhân giữa các giá trị logic, sau đó in ra tổng. Kết quả cuối cùng là 3.

### 2.4 Empty Object

In [4]:
x = None 
print(x)

None


Gán giá trị đặc biệt None (đại diện cho “không có giá trị” hoặc “rỗng”) cho biến x. In ra giá trị của x, kết quả hiển thị là None.

In [5]:
print(type(x))

<class 'NoneType'>


In ra kiểu dữ liệu của biến x. Vì x đang được gán giá trị đặc biệt None, nên kiểu dữ liệu của nó là class 'NoneType'.

In [7]:
x = 1
y = None
print(x is None)

False


Đoạn lệnh kiểm tra xem x có phải là None không. Vì x = 1 nên kết quả là False.

In [8]:
print(y is None)

True


Đoạn lệnh kiểm tra xem y có phải là None không. Vì y = None nên kết quả là True.

### 2.5 Dates and Times

#### 2.5.1 Module Datetime

##### 2.5.1.1 Date 

In [19]:
from datetime import date
debut = date(year = 2013, month = 4, day = 23) 
print(debut)

2013-04-23


Lệnh này tạo một đối tượng ngày (date) với giá trị 23/04/2013 bằng cách chỉ định năm, tháng, ngày. Sau đó in ra giá trị của biến debut.

In [11]:
print(type(debut))

<class 'datetime.date'>


Lệnh này in ra kiểu dữ liệu của biến debut. kết quả cho biết debut là một đối tượng ngày thuộc lớp datetime.date.

In [12]:
print(debut.year) # Extract the year

2013


Lệnh này truy xuất năm từ đối tượng ngày debut.

In [13]:
print(debut.month) # Extract the month

4


Lệnh này truy xuất tháng từ đối tượng ngày debut.

In [14]:
print(debut.day) # Extract the day

23


Lệnh này truy xuất ngày từ đối tượng ngày debut.

##### 2.5.1.1.1 ctime()

In [16]:
print(debut.ctime())

Tue Apr 23 00:00:00 2013


Lệnh này chuyển đối tượng ngày debut thành một chuỗi ký tự dạng c-time (chuẩn định dạng ngày giờ kiểu C).

##### 2.5.1.1.2 weekday()

In [17]:
print(debut.weekday())

1


Lệnh print(debut.weekday()) sẽ trả về thứ trong tuần của đối tượng ngày debut, nhưng theo quy ước: 0 → Thứ Hai, 1 → Thứ Ba, 2 → Thứ Tư, 3 → Thứ Năm, 4 → Thứ Sáu, 5 → Thứ Bảy, 6 → Chủ Nhật.

##### 2.5.1.1.3 isoweekday()

In [18]:
print(debut.isoweekday())

2


Lệnh print(debut.isoweekday()) cũng trả về thứ trong tuần, nhưng theo ISO 8601: 1 → Thứ Hai, 2 → Thứ Ba, 3 → Thứ Tư, 4 → Thứ Năm, 5 → Thứ Sáu, 6 → Thứ Bảy, 7 → Chủ Nhật.

##### 2.5.1.1.4 toordinal()

In [19]:
print(debut.toordinal())

734981


Lệnh print(debut.toordinal()) sẽ trả về số nguyên thứ tự của ngày tính từ ngày 1/1/0001 (ngày đầu tiên trong lịch Gregorian) với giá trị 1. Với debut = date(2013, 4, 23), Python sẽ tính xem ngày này là ngày thứ bao nhiêu kể từ 01/01/0001.

##### 2.5.1.1.5 isoformat()

In [20]:
print(debut.isoformat())

2013-04-23


Lệnh print(debut.isoformat()) sẽ in ra ngày theo định dạng chuẩn ISO 8601

##### 2.5.1.1.6 isocalendar()

In [21]:
print(debut.isocalendar())

datetime.IsoCalendarDate(year=2013, week=17, weekday=2)


Lệnh print(debut.isocalendar()) sẽ trả về một tuple gồm 3 giá trị.

##### 2.5.1.1.7 replace()

In [23]:
x = debut.replace(year=2014) 
y = debut.replace(month=5)
z = debut.replace(day=24)
print(x, y,z)

2014-04-23 2013-05-23 2013-04-24


replace() sẽ tạo một đối tượng ngày mới với các giá trị được thay đổi mà vẫn giữ nguyên các thành phần khác.

In [24]:
print(debut)

2013-04-23


Dù trước đó dùng replace() để tạo x, y, z nhưng debut vẫn giữ nguyên giá trị gốc vì replace() không thay đổi trực tiếp đối tượng ban đầu.

In [25]:
x = debut.replace(day=24, month=5) 
print(x)

2013-05-24


replace() trong đối tượng date dùng để tạo ra một ngày mới bằng cách giữ nguyên các phần khác và thay thế những thành phần được chỉ định (năm, tháng, ngày). Nó không làm thay đổi ngày ban đầu, mà trả về một đối tượng mới.

##### 2.5.1.1.8 strftime()

In [26]:
print(debut.strftime("%d-%m-%Y"))

23-04-2013


Lệnh này chuyển đổi giá trị ngày debut thành chuỗi theo định dạng ngày-tháng-năm, giúp hiển thị ngày ở dạng văn bản dễ đọc hơn thay vì dạng đối tượng date.

In [27]:
print(debut.strftime("%d/%m/%Y"))

23/04/2013


Định dạng và hiển thị ngày tháng của biến debut dưới dạng ngày/tháng/năm với dấu gạch chéo / làm ký tự ngăn cách.

In [28]:
print(debut.strftime("Jour : %d, Mois : %m, Annee : %Y"))

Jour : 23, Mois : 04, Annee : 2013


Hiển thị ngày debut dưới dạng chuỗi với nhãn rõ ràng cho ngày, tháng, năm, giúp đọc dễ hiểu hơn.

##### 2.5.1.2 Time 

In [32]:
from datetime import time
moment = time(hour = 23, minute = 4, second = 59)
print(moment)

23:04:59


Tạo và hiển thị một thời điểm trong ngày (giờ, phút, giây) bằng đối tượng time trong Python.

In [33]:
print(type(moment))

<class 'datetime.time'>


Xác định rằng moment là một đối tượng thời gian thuộc lớp datetime.time.

In [34]:
moment = time(hour = 23, minute = 4, second = 59, microsecond = 230) 
print(moment)

23:04:59.000230


Tạo và hiển thị một thời điểm trong ngày với giờ, phút, giây và microgiây bằng đối tượng time trong Python.

In [35]:
print(type(moment))

<class 'datetime.time'>


Xác định rằng moment là một đối tượng thời gian thuộc lớp datetime.time, có thể chứa giờ, phút, giây và microgiây.

In [36]:
print(moment.hour) # Extract the hour

23


Lấy và in giá trị giờ của đối tượng time.

In [37]:
print(moment.minute) # Extract the minute

4


Lấy và in giá trị phút của đối tượng time.

In [38]:
print(moment.second) # Extract the second

59


Lấy và in giá trị giây của đối tượng time.

In [40]:
print(moment.microsecond) # Extract the microsecond

230


Lấy và in giá trị microgiây của đối tượng time.

##### 2.5.1.3 Datetime

In [2]:
from datetime import datetime
x = datetime(year = 2013, month = 4, day = 23,
 hour = 23, minute = 4, second = 59) 
print(x)

2013-04-23 23:04:59


In [3]:
print(type(x))

<class 'datetime.datetime'>


Xác định rằng x là một đối tượng ngày giờ đầy đủ thuộc lớp datetime.datetime.

##### 2.5.1.3.1 today() et now()

In [4]:
print(x.today())

2025-08-14 07:36:39.197114


Lệnh print(x.today()) sẽ in ra ngày và giờ hiện tại tại thời điểm chạy lệnh.

In [5]:
print(datetime.today())

2025-08-14 07:37:15.129385


Lệnh này in ra ngày và giờ hiện tại tại thời điểm chạy lệnh, sử dụng phương thức today() của lớp datetime.

##### 2.5.1.3.2 timestamp()

In [6]:
print(x.timestamp())

1366758299.0


Lệnh print(x.timestamp()) trả về giá trị thời gian dạng số thực (timestamp) tương ứng với đối tượng datetime x.

##### 2.5.1.3.3 date()

In [7]:
x_date = x.date()
print(x_date)

2013-04-23


Tách phần ngày từ một datetime để làm việc chỉ với ngày, không bao gồm giờ phút giây và in kết quả ra màn hình.

In [8]:
print(type(x_date))

<class 'datetime.date'>


Xác định rằng x_date là một đối tượng ngày (date) riêng, chỉ chứa thông tin năm, tháng, ngày, không có giờ phút giây.

##### 2.5.1.3.4 time()

In [9]:
x_time = x.time() 
print(x_time)

23:04:59


Tách phần thời gian từ một datetime để làm việc chỉ với giờ, phút, giây, không bao gồm ngày tháng.

In [10]:
print(type(x_time))

<class 'datetime.time'>


Xác định rằng x_time là một đối tượng thời gian (time) riêng, chỉ chứa thông tin giờ, phút, giây, không có ngày tháng.

##### 2.5.1.4 Timedelta

In [11]:
from datetime import timedelta
duree = timedelta(days = 1, seconds = 30) 
duree

datetime.timedelta(days=1, seconds=30)

Tạo và biểu diễn một khoảng thời gian (1 ngày và 30 giây) trong Python bằng đối tượng timedelta.

In [14]:
duree. days

1

Lấy và in ra số ngày của khoảng thời gian duree.

In [15]:
duree = timedelta(days = 1, seconds = 30, hours = 20) 
duree.total_seconds()

158430.0

Chuyển đổi toàn bộ khoảng thời gian timedelta thành số giây tổng cộng để thuận tiện cho tính toán.

##### 2.5.1.4.1 Time Between Two Objects date or datetime.

In [21]:
from datetime import timedelta 
beginning = date(2018, 1, 1) 
end = date(2018, 1, 2)
nb_days = end-beginning 
print(type(nb_days))

<class 'datetime.timedelta'>


Minh họa cách tính khoảng thời gian giữa hai mốc ngày (hoặc thời điểm) trong Python, và cho thấy phép trừ giữa hai đối tượng date sẽ trả về một đối tượng timedelta biểu diễn chênh lệch thời gian đó.

In [23]:
print(nb_days)

1 day, 0:00:00


Nghĩa là chênh lệch giữa 2018-01-02 và 2018-01-01 là 1 ngày, không có thêm giờ, phút, hay giây.

In [25]:
beginning = datetime(2018, 1, 1, 12, 26, 30, 230) 
end = datetime(2018, 1, 2, 11, 14, 31)
duration = end-beginning
print(type(duration))

<class 'datetime.timedelta'>


beginning và end là hai đối tượng datetime có cả ngày và giờ, phút, giây, micro giây. Phép trừ end - beginning trả về một đối tượng timedelta, thể hiện khoảng thời gian giữa hai mốc này. Khi print(type(duration)) sẽ in ra kết quả trả về màn hình như trên.

In [26]:
print(duration)

22:48:00.999770


Duration biểu diễn khoảng thời gian 22 giờ 48 phút và 0.999770 giây giữa hai thời điểm đó.

In [28]:
beginning = date(2021, 2,28) 
end = date(2021, 3, 1) 
duration = end - beginning 
duration

datetime.timedelta(days=1)

Đoạn code minh họa rằng khi trừ hai đối tượng date trong Python, kết quả là một đối tượng timedelta biểu diễn khoảng cách thời gian giữa chúng (ở đây là 1 ngày), và khoảng cách này được tính theo lịch, không liên quan đến giờ phút giây.

In [30]:
beginning_leap = date(2020, 2,28)
end_leap = date(2020, 3, 1)
beginning_leap = end_leap - beginning_leap 
beginning_leap

datetime.timedelta(days=2)

Khoảng cách giữa hai ngày là 2 ngày do năm 2020 có thêm ngày 29/02.

In [32]:
debut = datetime(2018, 12, 31, 23, 59, 59) 
print(debut + timedelta(seconds = 1))

2019-01-01 00:00:00


Chỉ cần cộng thêm 1 giây, thời gian đã nhảy sang ngày 1 tháng 1 năm 2019, 0 giờ 0 phút 0 giây.

#### 2.5.2 pytz Module

#### 2.5.3 Exercises

1. Using the appropriate function, store the date of August 29, 2019 in an object called d then display the type of the object.

In [35]:
from datetime import date
d = date(2019, 8, 29)
print(type(d))

<class 'datetime.date'>


2. Using the appropriate function, display the current date.

In [36]:
from datetime import date
today = date.today()
print(today)

2025-08-14


3. Store the next date in an object named d2 : “2019-08-29 20:30:56”. Then,
display in the console with the print() function the year, minute and second attributes of d2.

In [37]:
from datetime import datetime
d2 = datetime(2019, 8, 29, 20, 30, 56)
print(d2.year)  
print(d2.minute)
print(d2.second)

2019
30
56


4. Add 2 days, 3 hours and 4 minutes to d2, and store the result in an object called d3.

In [38]:
d3 = d2 + timedelta(days=2, hours=3, minutes=4)
print(d3)

2019-08-31 23:34:56


5. Display the difference in seconds between d3 and d2.

In [39]:
difference = d3 - d2
print(difference.total_seconds())

183840.0


6. From the object d2, display the date of d2 as a string so that it follows the following syntax: “Month Day, Year”, with “Month” the name of the month (August), “Day” the two-digit day number (29) and “Year” the year of the date (2019).

In [40]:
formatted_date = d2.strftime("%B %d, %Y")
print(formatted_date)

August 29, 2019


# Kết thúc Chapter 2